File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    bridge_queue = deque([0] * bridge_length)
    total_weight_on_bridge=0
    time=0
    truck_index=0
    
    while truck_index < len(truck_weights):
        time += 1
        total_weight_on_bridge -= bridge_queue.popleft()
        
        if truck_index < len(truck_weights) and total_weight_on_bridge + truck_weights[truck_index] <= weight:
            bridge_queue.append(truck_weights[truck_index])
            total_weight_on_bridge += truck_weights[truck_index]
            truck_index += 1
        else:
            bridge_queue.append(0)
    
    time += bridge_length
    
    return time

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    let time = 0;
    let bridge = Array(bridge_length).fill(0);
    let bridgeWeight = 0;

    while (truck_weights.length > 0 || bridgeWeight > 0) {
        time++;
        bridgeWeight -= bridge.shift();

        if (truck_weights.length > 0 && bridgeWeight + truck_weights[0] <= weight) {
            const nextTruck = truck_weights.shift();
            bridge.push(nextTruck);
            bridgeWeight += nextTruck;
        } else {
            bridge.push(0);
        }
    }

    return time;
}